### PR TITLE
Web API documentation

### PR DESCRIPTION
**Pull request by lewish**
Auto generated API documentation from the web API swagger spec.

image

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -28,11 +28,13 @@
     "@types/rimraf": "^2.0.2",
     "@types/semver": "^6.2.0",
     "@types/stack-trace": "^0.0.29",
+    "@types/swagger-schema-official": "^2.0.20",
     "@types/yargs": "^11.1.1",
     "@zeit/next-css": "1.0.1",
     "@zeit/next-mdx": "^1.2.0",
     "@zeit/next-typescript": "^1.1.1",
     "algoliasearch": "^3.32.1",
+    "axios": "^0.19.0",
```
Comment (GJMcGowan) [inline package.json]: I'm surprised that this addition didn't change the `yarn.lock`

File: docs/components/swagger.css [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+.propertyName {
+  max-width: 300px;
+  width: 50%;
+  display: inline-block;
+}
+
+.propertyDescription {
+  padding: 5px;
+  padding-left: 300px;
```
Comment (GJMcGowan) [inline docs/components/swagger.css]: This is a very weird thing to do. Instead, consider just merging the `code` and `propertyDescription` elements into one and then having the page flow normally.

File: docs/global.css [hunk 3]
```diff
@@ -98,13 +98,24 @@
   color: #2f2d44;
 }
 
+:global body h5 {
```
Comment (GJMcGowan) [inline docs/global.css]: This whole file is quite repetitive.

Can you have a shared block for all the `h*`s which contains:
```
  font-family: "Poppins";
  font-weight: 500;
  letter-spacing: -0.02em;
  color: #2f2d44;
```

and then apply custom styling for each of them

File: docs/global.css [hunk 4]
```diff
@@ -98,13 +98,24 @@
...
+  color: colorsTextMain;
```
Comment (GJMcGowan) [inline docs/global.css]: Also, this color is overridden at the bottom of this block, so it's pointless.

File: docs/components/swagger.tsx [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+      .reduce((acc, curr) => [...acc, ...curr], []);
+
+    return (
+      <div className={styles.container}>
+        <div className={styles.sidebar} />
+        <div className={styles.mainContent}>
+          <div className={styles.titleBlock}>
+            <h1>Dataform Web API</h1>
+            {allOperations.map((operation, i) => (
+              <Operation key={i} apiHost={apiHost} operation={operation} />
```
Comment (GJMcGowan) [inline docs/components/swagger.tsx]: [Don't use indexes as keys](https://medium.com/@robinpokorny/index-as-a-key-is-an-anti-pattern-e0349aece318)

File: docs/components/swagger.tsx [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+            ))}
+            {Object.keys(spec.definitions).map((name, i) => (
+              <Definition key={i} name={name} schema={spec.definitions[name]} />
```
Comment (GJMcGowan) [inline docs/components/swagger.tsx]: Also here

Comment (GJMcGowan) [reply docs/components/swagger.tsx]: (And a lot of other places)

File: package.json (hunk 1 above)
Comment (lewish) [reply package.json]: It's already in there, just adding it as an explicit dependency now.

File: docs/components/swagger.tsx (hunk 5 above)
Comment (lewish) [reply docs/components/swagger.tsx]: As a general rule yes - however this content is completely static, it never changes as the data model will never change unless the user refreshes, at which point it doesn't matter.

Comment (lewish) [reply docs/components/swagger.tsx]: OK OK I changed them

File: docs/components/swagger.css (hunk 2 above)
Comment (lewish) [reply docs/components/swagger.css]: Moved to flex

**Comment by lewish** [comment]
> What's up with the left hand sidebar being empty?

This is on it's own at the moment as we don't want to publish it (Well, it's open-source, but just not promoted). I'm shortly about to make a big change to how the sidebar works that will somewhat fix this.

File: docs/components/swagger.css [hunk 7]
```diff
@@ -39,23 +39,24 @@
...
 
 .properties {
-  
 }
 
 .propertyName {
   max-width: 300px;
-  width: 50%;
-  display: inline-block;
+  min-width: 300px;
```
Comment (GJMcGowan) [inline docs/components/swagger.css]: Perhaps just `width` if the min and max are the same.

File: docs/global.css [hunk 8]
```diff
@@ -55,58 +55,49 @@
...
+  letter-spacing: -0.03em;
+  color: #2f2d44;
+}
+
+:global body h1 {
   font-weight: 500;
   letter-spacing: -0.03em;
   line-height: 58px;
   font-size: 31px;
-  margin: 0px;
```
Comment (GJMcGowan) [inline docs/global.css]: I think you forgot to put the `margin: 0` on the default styles.

File: docs/global.css [hunk 9]
```diff
@@ -55,58 +55,49 @@
...
+:global h4,
+:global h5 {
   font-family: "Poppins";
+  letter-spacing: -0.03em;
+  color: #2f2d44;
+}
+
+:global body h1 {
   font-weight: 500;
   letter-spacing: -0.03em;
```
Comment (GJMcGowan) [inline docs/global.css]: `letter-spacing` can probably just be the one in the default styles as well.

Comment (GJMcGowan) [reply docs/global.css]: Same with `font-weight`

File: docs/global.css (hunk 8 above)
Comment (lewish) [reply docs/global.css]: Didn't actually ever need this

Comment (GJMcGowan) [reply docs/global.css]: `h*` elements have a default margin which this overrides. Your docs may look weird with more margin. Your choice though 😄

**Pull request merged**